### ISSUE TITLE
`unused_assignments` when writing to variable and then reading from its raw pointer

### ISSUE THREAD

**Issue by Giftzwerg02** [rust-lang/rust#150656]
### Code

```Rust
let mut a = 0;
let a_ptr = &mut a as *mut _;

a = 3; // <-- produces warning for this assignment

println!("a: {}", unsafe { *a_ptr });
```

### Current output

```Shell
warning: value assigned to `a` is never read
  --> src/main.rs:43:5
   |
43 |     a = 3;
   |     ^^^^^
   |
   = help: maybe it is overwritten before being read?
   = note: `#[warn(unused_assignments)]` (part of `#[warn(unused)]`) on by default

warning: 1 warning emitted
```

### Desired output

```Shell
Ideally no error at all (?)
```

### Rationale and extra context

I first got this warning when using the libc-crate, which requires passing raw pointers, e.g.:
```rust
fn main() {
    use libc;
    use std::time::Duration;

    let timeout = Duration::from_secs(1);
    let mut ts = libc::timespec {
        tv_sec: 0,
        tv_nsec: 0,
    };

    let ts_ptr = &mut ts as *mut _;

    unsafe { libc::clock_gettime(libc::CLOCK_REALTIME, ts_ptr) };

    // these writes cause a "value assigned is never read" warning
    ts.tv_sec += timeout.as_secs() as i64;
    ts.tv_nsec += timeout.subsec_nanos() as i64;

    let mut rc = 0;
    while /* exit-condition && */ rc == 0 {
        // ts_ptr used again here
        rc = unsafe { libc::pthread_cond_timedwait(todo!("cond"), todo!("lock"), ts_ptr) }
    }
}
```

### Other cases

```Rust

```

### Rust Version

```Shell
$ rustc --version --verbose
rustc 1.92.0 (ded5c06cf 2025-12-08)
binary: rustc
commit-hash: ded5c06cf21d2b93bffd5d884aa6e96934ee4234
commit-date: 2025-12-08
host: x86_64-unknown-linux-gnu
release: 1.92.0
LLVM version: 21.1.3
```

**Comment by ver-nyan** [comment]
I'm also running into the same issue in https://github.com/libbpf/libbpf-rs/pull/1352.
Slightly modified snippet from link:
```rust
let mut linkinfo: libbpf_sys::bpf_iter_link_info = unsafe { mem::zeroed() };

let (_linkinfo_ptr, _linkinfo_len) = match &opts {
    IterOpts::None => (ptr::null_mut(), 0),
    IterOpts::Map(_) | IterOpts::Cgroup(_) => (
        &raw mut linkinfo,
        size_of::<libbpf_sys::bpf_iter_link_info>() as _,
    ),
};

match opts {
    IterOpts::Map(map_opts) => {
        linkinfo.map.map_fd = map_opts.fd.as_raw_fd() as _;                                 // <-- unused warning
    }
    IterOpts::Cgroup(cgroup_opts) => {
        linkinfo.cgroup.order = cgroup_opts.order as libbpf_sys::bpf_cgroup_iter_order;     // <-- unused warning
        linkinfo.cgroup.cgroup_fd = cgroup_opts.fd.as_raw_fd() as _;                        // <-- unused warning
    }
    _ => {}
}
```
also ignore how `IterOpts::None` looks, api is still wip :)

I have my pointer initialized early to satisfy the borrow checker when using `opts` twice.

But it's not a big deal since I can probably workaround to satisfy warning. just wanted to point out that I also ran into this too.
